# Instructions

Tonight you're going to practice working in Pandas. You'll walk through instantiating a `DataFrame`, reading data into it, looking at and examining that data, and then playing with it. We'll be using the data in the `data` folder located within this directory (it's the same wine data that we looked at during lecture). Typically, we use IPython notebooks like this for a very specific set of things - presentations and EDA. Tonight, as we'll be playing around with `Pandas`, much of what we'll be doing is considered EDA. Therefore, by using a notebook, we'll get a tighter feedback loop with our work than we would trying to write a script. But, in general, **we do not use IPython notebooks for development**. 

Below, we've put a set of questions and then a cell for you to work on answers. However, feel free to add additional cells if you'd like. Often it will make sense to use more than one cell for your answers. 

# Assignment Questions 

### Part 1 - The Basics of DataFrames

Let's start off by following the general workflow that we use when moving data into a DataFrame: 

    * Importing Pandas
    * Reading data into the DataFrame
    * Getting a general sense of the data

Your tasks, right now:

1. Import pandas
2. Read the wine data into a DataFrame. 
3. Use the `attributes` and `methods` available on DataFrames to answer the following questions: 
    * How many rows and columns are in the DataFrame?
    * What data type is in each column?
    * Are all of the variables continuous, or are any categorical?
    * How many non-null values are in each column?
    * What are the min, mean, max, median for all numeric columns?

In [2]:
import pandas as pd
import numpy as np
red_df = pd.read_csv('./data/winequality-red.csv', sep=';')
red_df.head(10)

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0            7.4              0.70         0.00             1.9      0.076   
1            7.8              0.88         0.00             2.6      0.098   
2            7.8              0.76         0.04             2.3      0.092   
3           11.2              0.28         0.56             1.9      0.075   
4            7.4              0.70         0.00             1.9      0.076   
5            7.4              0.66         0.00             1.8      0.075   
6            7.9              0.60         0.06             1.6      0.069   
7            7.3              0.65         0.00             1.2      0.065   
8            7.8              0.58         0.02             2.0      0.073   
9            7.5              0.50         0.36             6.1      0.071   

   free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
0                 11.0                  34.0   0.9978  3.51       0.56   
1                 25.0                  67.0   0.9968  3.20       0.68   
2                 15.0                  54.0   0.9970  3.26       0.65   
3                 17.0                  60.0   0.9980  3.16       0.58   
4                 11.0                  34.0   0.9978  3.51       0.56   
5                 13.0                  40.0   0.9978  3.51       0.56   
6                 15.0                  59.0   0.9964  3.30       0.46   
7                 15.0                  21.0   0.9946  3.39       0.47   
8                  9.0                  18.0   0.9968  3.36       0.57   
9                 17.0                 102.0   0.9978  3.35       0.80   

   alcohol  quality  
0      9.4        5  
1      9.8        5  
2      9.8        5  
3      9.8        6  
4      9.4        5  
5      9.4        5  
6      9.4        5  
7     10.0        7  
8      9.5        7  
9     10.5        5

 * What data type is in each column?

In [27]:
red_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1599 entries, 0 to 1598
Data columns (total 12 columns):
fixed acidity           1599 non-null float64
volatile acidity        1599 non-null float64
citric acid             1599 non-null float64
residual sugar          1599 non-null float64
chlorides               1599 non-null float64
free sulfur dioxide     1599 non-null float64
total sulfur dioxide    1599 non-null float64
density                 1599 non-null float64
pH                      1599 non-null float64
sulphates               1599 non-null float64
alcohol                 1599 non-null float64
quality                 1599 non-null int64
dtypes: float64(11), int64(1)
memory usage: 150.0 KB


In [4]:
red_df.shape

(1599, 12)

In [24]:
red_df.describe()

fixed acidity  volatile acidity  citric acid  residual sugar  \
count    1599.000000       1599.000000  1599.000000     1599.000000   
mean        8.319637          0.527821     0.270976        2.538806   
std         1.741096          0.179060     0.194801        1.409928   
min         4.600000          0.120000     0.000000        0.900000   
25%         7.100000          0.390000     0.090000        1.900000   
50%         7.900000          0.520000     0.260000        2.200000   
75%         9.200000          0.640000     0.420000        2.600000   
max        15.900000          1.580000     1.000000       15.500000   

         chlorides  free sulfur dioxide  total sulfur dioxide      density  \
count  1599.000000          1599.000000           1599.000000  1599.000000   
mean      0.087467            15.874922             46.467792     0.996747   
std       0.047065            10.460157             32.895324     0.001887   
min       0.012000             1.000000              6.000000     0.990070   
25%       0.070000             7.000000             22.000000     0.995600   
50%       0.079000            14.000000             38.000000     0.996750   
75%       0.090000            21.000000             62.000000     0.997835   
max       0.611000            72.000000            289.000000     1.003690   

                pH    sulphates      alcohol      quality  
count  1599.000000  1599.000000  1599.000000  1599.000000  
mean      3.311113     0.658149    10.422983     5.636023  
std       0.154386     0.169507     1.065668     0.807569  
min       2.740000     0.330000     8.400000     3.000000  
25%       3.210000     0.550000     9.500000     5.000000  
50%       3.310000     0.620000    10.200000     6.000000  
75%       3.400000     0.730000    11.100000     6.000000  
max       4.010000     2.000000    14.900000     8.000000

* How many non-null values are in each column?

In [30]:
red_df[red_df.isnull().any(axis=1)].count()

fixed acidity           0
volatile acidity        0
citric acid             0
residual sugar          0
chlorides               0
free sulfur dioxide     0
total sulfur dioxide    0
density                 0
pH                      0
sulphates               0
alcohol                 0
quality                 0
dtype: int64

In [37]:
## check to see how to count null values for each column in a more readable way...

In [31]:
copy_df = red_df
copy_df['dummy'] = np.nan

In [34]:
copy_df.head()

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0            7.4              0.70         0.00             1.9      0.076   
1            7.8              0.88         0.00             2.6      0.098   
2            7.8              0.76         0.04             2.3      0.092   
3           11.2              0.28         0.56             1.9      0.075   
4            7.4              0.70         0.00             1.9      0.076   

   free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
0                 11.0                  34.0   0.9978  3.51       0.56   
1                 25.0                  67.0   0.9968  3.20       0.68   
2                 15.0                  54.0   0.9970  3.26       0.65   
3                 17.0                  60.0   0.9980  3.16       0.58   
4                 11.0                  34.0   0.9978  3.51       0.56   

   alcohol  quality  dummy  
0      9.4        5    NaN  
1      9.8        5    NaN  
2      9.8        5    NaN  
3      9.8        6    NaN  
4      9.4        5    NaN

In [45]:
copy_df.isnull().sum(axis=0)

fixed acidity              0
volatile acidity           0
citric acid                0
residual sugar             0
chlorides                  0
free sulfur dioxide        0
total sulfur dioxide       0
density                    0
pH                         0
sulphates                  0
alcohol                    0
quality                    0
dummy                   1599
dtype: int64

### Part 2 - Practice with Grabbing Data

Let's now get some practice with grabbing certain parts of the data. If you'd like some extra practice, try answering each of the questions in more than one way (because remember, we can often grab our data in a couple of different ways). 

1. Grab the first 10 rows of the `chlorides` column. 

In [5]:
red_df.loc[ 0:10, 'chlorides']

0     0.076
1     0.098
2     0.092
3     0.075
4     0.076
5     0.075
6     0.069
7     0.065
8     0.073
9     0.071
10    0.097
Name: chlorides, dtype: float64

In [9]:
red_df.iloc[ -10:  , 4]

1589    0.073
1590    0.077
1591    0.089
1592    0.076
1593    0.068
1594    0.090
1595    0.062
1596    0.076
1597    0.075
1598    0.067
Name: chlorides, dtype: float64

#### 2 -  Grab the last 10 rows of the `chlorides` column.

In [7]:
red_df['chlorides'].tail(10)

1589    0.073
1590    0.077
1591    0.089
1592    0.076
1593    0.068
1594    0.090
1595    0.062
1596    0.076
1597    0.075
1598    0.067
Name: chlorides, dtype: float64

####  3. Grab indices 264-282 of the `chlorides` **and** `density` columns. 

In [10]:
red_df.loc[264:282, ['density', 'chlorides']]

density  chlorides
264  0.99990      0.064
265  0.99680      0.071
266  1.00025      0.096
267  0.99730      0.078
268  0.99870      0.077
269  0.99960      0.104
270  0.99650      0.087
271  0.99960      0.104
272  0.99935      0.071
273  0.99735      0.076
274  0.99915      0.088
275  0.99650      0.087
276  0.99870      0.077
277  0.99960      0.104
278  0.99760      0.073
279  0.99910      0.087
280  0.99860      0.071
281  0.99720      0.358
282  0.99710      0.067

#### 4. Grab all rows where the `chlorides` value is less than 0.10. 

In [66]:
red_df.query('chlorides < .10').sort_values('chlorides')

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
837             6.7             0.280         0.28            2.40      0.012   
836             6.7             0.280         0.28            2.40      0.012   
695             5.1             0.470         0.02            1.30      0.034   
1571            6.4             0.380         0.14            2.20      0.038   
797             9.3             0.370         0.44            1.60      0.038   
210             9.7             0.530         0.60            2.00      0.039   
1490            7.1             0.220         0.49            1.80      0.039   
861             5.8             1.010         0.66            2.00      0.039   
688             7.7             0.660         0.04            1.60      0.039   
806             8.4             0.250         0.39            2.00      0.041   
1321            5.0             0.740         0.00            1.20      0.041   
1316            5.4             0.740         0.00            1.20      0.041   
1178            5.6             0.915         0.00            2.10      0.041   
1432            5.9             0.440         0.00            1.60      0.042   
1483            8.1             0.330         0.44            1.50      0.042   
1157            5.1             0.510         0.18            2.10      0.042   
1142            6.9             0.450         0.11            2.40      0.043   
1269            5.5             0.490         0.03            1.80      0.044   
1219            9.0             0.390         0.40            1.30      0.044   
1151            6.1             0.580         0.23            2.50      0.044   
1228            5.1             0.420         0.00            1.80      0.044   
802             5.1             0.585         0.00            1.70      0.044   
1131            5.9             0.190         0.21            1.70      0.045   
426             6.4             0.670         0.08            2.10      0.045   
390             5.6             0.850         0.05            1.40      0.045   
94              5.0             1.020         0.04            1.40      0.045   
1426            7.7             0.230         0.37            1.80      0.046   
1044            6.4             0.390         0.33            3.30      0.046   
198             5.4             0.835         0.08            1.20      0.046   
1114            5.0             0.400         0.50            4.30      0.046   
...             ...               ...          ...             ...        ...   
690             7.4             1.185         0.00            4.25      0.097   
238             7.2             0.630         0.00            1.90      0.097   
988             7.7             0.390         0.12            1.70      0.097   
752             7.6             0.540         0.13            2.50      0.097   
990             7.7             0.390         0.12            1.70      0.097   
819             9.1             0.660         0.15            3.20      0.097   
10              6.7             0.580         0.08            1.80      0.097   
119             7.0             0.690         0.08            1.80      0.097   
237             7.2             0.645         0.00            1.90      0.097   
628             8.7             0.540         0.26            2.50      0.097   
531            11.9             0.380         0.49            2.70      0.098   
1               7.8             0.880         0.00            2.60      0.098   
1023            8.2             0.320         0.42            2.30      0.098   
1074            7.5             0.770         0.20            8.10      0.098   
532            11.9             0.380         0.49            2.70      0.098   
465            10.0             0.290         0.40            2.90      0.098   
627             8.8             0.600         0.29            2.20      0.098   
463             8.1             0.660

#### 5. Now grab all the rows where the `chlorides` value is greater than the column's mean (try **not** to use a hard-coded value for the mean, but instead a method).

In [74]:
chlorides_mean = red_df['chlorides'].mean()
red_df.query('chlorides < @chlorides_mean')

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0               7.4             0.700         0.00             1.9      0.076   
3              11.2             0.280         0.56             1.9      0.075   
4               7.4             0.700         0.00             1.9      0.076   
5               7.4             0.660         0.00             1.8      0.075   
6               7.9             0.600         0.06             1.6      0.069   
7               7.3             0.650         0.00             1.2      0.065   
8               7.8             0.580         0.02             2.0      0.073   
9               7.5             0.500         0.36             6.1      0.071   
11              7.5             0.500         0.36             6.1      0.071   
18              7.4             0.590         0.08             4.4      0.086   
20              8.9             0.220         0.48             1.8      0.077   
21              7.6             0.390         0.31             2.3      0.082   
23              8.5             0.490         0.11             2.3      0.084   
24              6.9             0.400         0.14             2.4      0.085   
25              6.3             0.390         0.16             1.4      0.080   
26              7.6             0.410         0.24             1.8      0.080   
28              7.1             0.710         0.00             1.9      0.080   
29              7.8             0.645         0.00             2.0      0.082   
32              8.3             0.655         0.12             2.3      0.083   
33              6.9             0.605         0.12            10.7      0.073   
35              7.8             0.645         0.00             5.5      0.086   
36              7.8             0.600         0.14             2.4      0.086   
37              8.1             0.380         0.28             2.1      0.066   
39              7.3             0.450         0.36             5.9      0.074   
40              7.3             0.450         0.36             5.9      0.074   
43              8.1             0.660         0.22             2.2      0.069   
44              6.8             0.670         0.02             1.8      0.050   
45              4.6             0.520         0.15             2.1      0.054   
48              6.4             0.400         0.23             1.6      0.066   
49              5.6             0.310         0.37             1.4      0.074   
...             ...               ...          ...             ...        ...   
1565            6.7             0.670         0.02             1.9      0.061   
1566            6.7             0.160         0.64             2.1      0.059   
1567            7.2             0.695         0.13             2.0      0.076   
1568            7.0             0.560         0.13             1.6      0.077   
1569            6.2             0.510         0.14             1.9      0.056   
1571            6.4             0.380         0.14             2.2      0.038   
1572            7.3             0.690         0.32             2.2      0.069   
1573            6.0             0.580         0.20             2.4      0.075   
1574            5.6             0.310         0.78            13.9      0.074   
1575            7.5             0.520         0.40             2.2      0.060   
1576            8.0             0.300         0.63             1.6      0.081   
1577            6.2             0.700         0.15             5.1      0.076   
1579            6.2             0.560         0.09             1.7      0.053   
1580            7.4             0.350         0.33             2.4      0.068   
1581            6.2             0.560         0.09             1.7      0.053   
1582            6.1             0.715         0.10             2.6      0.053   
1583            6.2             0.460         0.29             2.1      0.074   
1584            6.7             0.320

#### 6. Grab all those rows where the `pH` is greater than 3.0 and less than 3.5. 

In [79]:
red_df.query('3.0 < pH < 3.5')

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
1               7.8             0.880         0.00             2.6      0.098   
2               7.8             0.760         0.04             2.3      0.092   
3              11.2             0.280         0.56             1.9      0.075   
6               7.9             0.600         0.06             1.6      0.069   
7               7.3             0.650         0.00             1.2      0.065   
8               7.8             0.580         0.02             2.0      0.073   
9               7.5             0.500         0.36             6.1      0.071   
10              6.7             0.580         0.08             1.8      0.097   
11              7.5             0.500         0.36             6.1      0.071   
13              7.8             0.610         0.29             1.6      0.114   
14              8.9             0.620         0.18             3.8      0.176   
15              8.9             0.620         0.19             3.9      0.170   
16              8.5             0.280         0.56             1.8      0.092   
17              8.1             0.560         0.28             1.7      0.368   
18              7.4             0.590         0.08             4.4      0.086   
19              7.9             0.320         0.51             1.8      0.341   
20              8.9             0.220         0.48             1.8      0.077   
22              7.9             0.430         0.21             1.6      0.106   
23              8.5             0.490         0.11             2.3      0.084   
24              6.9             0.400         0.14             2.4      0.085   
25              6.3             0.390         0.16             1.4      0.080   
26              7.6             0.410         0.24             1.8      0.080   
27              7.9             0.430         0.21             1.6      0.106   
28              7.1             0.710         0.00             1.9      0.080   
29              7.8             0.645         0.00             2.0      0.082   
30              6.7             0.675         0.07             2.4      0.089   
31              6.9             0.685         0.00             2.5      0.105   
32              8.3             0.655         0.12             2.3      0.083   
33              6.9             0.605         0.12            10.7      0.073   
34              5.2             0.320         0.25             1.8      0.103   
...             ...               ...          ...             ...        ...   
1560            7.8             0.600         0.26             2.0      0.080   
1561            7.8             0.600         0.26             2.0      0.080   
1562            7.2             0.695         0.13             2.0      0.076   
1563            7.2             0.695         0.13             2.0      0.076   
1564            7.2             0.695         0.13             2.0      0.076   
1565            6.7             0.670         0.02             1.9      0.061   
1566            6.7             0.160         0.64             2.1      0.059   
1567            7.2             0.695         0.13             2.0      0.076   
1568            7.0             0.560         0.13             1.6      0.077   
1569            6.2             0.510         0.14             1.9      0.056   
1570            6.4             0.360         0.53             2.2      0.230   
1571            6.4             0.380         0.14             2.2      0.038   
1572            7.3             0.690         0.32             2.2      0.069   
1574            5.6             0.310         0.78            13.9      0.074   
1575            7.5             0.520         0.40             2.2      0.060   
1576            8.0             0.300         0.63             1.6      0.081   
1578            6.8             0.670         0.15             1.8      0.118   
1580            7.4             0.350

#### 7. Further filter the results from 6 to grab only those rows that have a `residual sugar` less than 2.0. 

In [21]:
red_df = red_df.rename(columns={'residual sugar': 'residual_sugar'})
red_df.columns.values


array(['fixed acidity', 'volatile acidity', 'citric acid',
       'residual_sugar', 'chlorides', 'free sulfur dioxide',
       'total sulfur dioxide', 'density', 'pH', 'sulphates', 'alcohol',
       'quality'], dtype=object)

In [95]:
red_df.query('3.0 < pH < 3.5 and residual_sugar  < 2.0')

fixed acidity  volatile acidity  citric acid  residual_sugar  chlorides  \
3              11.2             0.280         0.56             1.9      0.075   
6               7.9             0.600         0.06             1.6      0.069   
7               7.3             0.650         0.00             1.2      0.065   
10              6.7             0.580         0.08             1.8      0.097   
13              7.8             0.610         0.29             1.6      0.114   
16              8.5             0.280         0.56             1.8      0.092   
17              8.1             0.560         0.28             1.7      0.368   
19              7.9             0.320         0.51             1.8      0.341   
20              8.9             0.220         0.48             1.8      0.077   
22              7.9             0.430         0.21             1.6      0.106   
25              6.3             0.390         0.16             1.4      0.080   
26              7.6             0.410         0.24             1.8      0.080   
27              7.9             0.430         0.21             1.6      0.106   
28              7.1             0.710         0.00             1.9      0.080   
34              5.2             0.320         0.25             1.8      0.103   
44              6.8             0.670         0.02             1.8      0.050   
47              8.7             0.290         0.52             1.6      0.113   
48              6.4             0.400         0.23             1.6      0.066   
49              5.6             0.310         0.37             1.4      0.074   
50              8.8             0.660         0.26             1.7      0.074   
61              7.7             0.690         0.49             1.8      0.115   
62              7.5             0.520         0.16             1.9      0.085   
66              7.5             0.520         0.11             1.5      0.079   
67              6.6             0.705         0.07             1.6      0.076   
69              8.0             0.705         0.05             1.9      0.074   
70              7.7             0.630         0.08             1.9      0.076   
72              7.7             0.690         0.22             1.9      0.084   
79              8.3             0.625         0.20             1.5      0.080   
80              6.2             0.450         0.20             1.6      0.069   
81              7.8             0.430         0.70             1.9      0.464   
...             ...               ...          ...             ...        ...   
1457            7.6             0.490         0.33             1.9      0.074   
1458            8.4             0.290         0.40             1.7      0.067   
1459            7.9             0.200         0.35             1.7      0.054   
1464            6.8             0.590         0.10             1.7      0.063   
1465            6.8             0.590         0.10             1.7      0.063   
1473            6.1             0.600         0.08             1.8      0.071   
1483            8.1             0.330         0.44             1.5      0.042   
1490            7.1             0.220         0.49             1.8      0.039   
1493            7.7             0.540         0.26             1.9      0.089   
1494            6.4             0.310         0.09             1.4      0.066   
1495            7.0             0.430         0.02             1.9      0.080   
1496            7.7             0.540         0.26             1.9      0.089   
1500            7.5             0.725         0.04             1.5      0.076   
1509            7.9             0.180         0.40             1.8      0.062   
1513            6.4             0.560         0.15             1.8      0.078   
1521            6.9             0.480         0.20             1.9      0.082   
1527            7.1             0.530         0.07             1.7      0.071   
1533            8.0             0.390

### Part 3 - More Practice

Let's move on to some more complicated things. Use your knowledge of `groupby`s, `sorting`, and the other things that you learned in lecture to answer the following. 

1. Get the average amount of `chlorides` for each `quality` value. 

In [16]:
red_df.groupby('quality').mean()['chlorides']

quality
3    0.122500
4    0.090679
5    0.092736
6    0.084956
7    0.076588
8    0.068444
Name: chlorides, dtype: float64

2 - For observations with a `pH` greater than 3.0 and less than 4.0, find the average `alcohol` value by `pH`. 

In [24]:
#red_df.columns.values
red_df.query('pH > 3.0 and pH < 4.0').groupby('pH').mean()

fixed acidity  volatile acidity  citric acid  residual_sugar  chlorides  \
pH                                                                              
3.01       9.960000          0.398000     0.640000        5.340000   0.136400   
3.02      10.962500          0.445625     0.501250        2.612500   0.119875   
3.03      10.683333          0.443333     0.543333        2.000000   0.199167   
3.04      10.270000          0.396000     0.522000        2.230000   0.144600   
3.05      11.337500          0.446250     0.518750        2.968750   0.086000   
3.06      11.160000          0.495500     0.569000        2.505000   0.168600   
3.07      12.236364          0.461818     0.480000        3.100000   0.088636   
3.08       9.272727          0.557273     0.440909        2.254545   0.118091   
3.09      10.290909          0.513636     0.430000        2.318182   0.115000   
3.10      10.315789          0.496579     0.412632        2.313158   0.083895   
3.11       9.900000          0.491111     0.424444        2.466667   0.128778   
3.12       9.730000          0.477250     0.387000        2.925000   0.086200   
3.13       9.953846          0.485385     0.400000        2.207692   0.120154   
3.14       9.995238          0.455714     0.359524        2.447619   0.082619   
3.15       9.679412          0.458824     0.394118        2.344118   0.089176   
3.16       9.500000          0.515139     0.354167        3.697222   0.105139   
3.17       9.581481          0.511296     0.396296        2.614815   0.101037   
3.18       9.923333          0.450167     0.403333        3.368333   0.096733   
3.19       9.424000          0.527200     0.325600        2.414000   0.100400   
3.20       9.289744          0.506410     0.363333        2.674359   0.084128   
3.21       8.802778          0.466667     0.357500        2.577778   0.093083   
3.22       9.320513          0.466795     0.351538        2.983333   0.089821   
3.23       8.912500          0.473281     0.342813        2.537500   0.081906   
3.24       9.034483          0.546034     0.317931        2.479310   0.088862   
3.25       8.707692          0.560962     0.323462        2.234615   0.092885   
3.26       8.513208          0.513491     0.315283        2.322642   0.083113   
3.27       8.514286          0.470714     0.341143        2.791429   0.081600   
3.28       8.880952          0.479167     0.318095        2.688095   0.090619   
3.29       8.113043          0.555109     0.245217        2.496739   0.076348   
3.30       8.384211          0.504123     0.288772        2.624561   0.085456   
...             ...               ...          ...             ...        ...   
3.46       7.022727          0.561136     0.164545        2.550000   0.079318   
3.47       7.380952          0.539048     0.140476        2.545238   0.078667   
3.48       7.642105          0.633947     0.142632        2.310526   0.075158   
3.49       6.860000          0.572500     0.155000        2.330000   0.071100   
3.50       7.035714          0.703571     0.134286        2.114286   0.086714   
3.51       7.146667          0.712333     0.064667        2.420000   0.076000   
3.52       6.772222          0.605278     0.141111        2.472222   0.071833   
3.53       7.147059          0.718235     0.123529        2.576471   0.074765   
3.54       6.587500          0.657813     0.139375        2.275000   0.068875   
3.55       6.725000          0.617500     0.078750        2.175000   0.074125   
3.56       6.009091          0.632727     0.084545        2.354545   0.083455   
3.57       6.080000          0.591000     0.121000        2.020000   0.068600   
3.58       6.680000          0.676500     0.108000        2.550000   0.079800   
3.59       6.400000          0.683125     0.101250        1.975000   0.080500   
3.60       6.800000          0.722857     0.071429        2.342857   0.079429   
3.61       6.762500          0.569375     0.095000        2.100000   0.078000   
3.62       6.125000          0.615000

3 -  For observations with an `alcohol` value between 9.25 and 9.5, find the highest amount of `residual sugar`. 

4 - Create a new column, called `total_acidity`, that is the sum of `fixed acidity` and `volatile acidity`. 

5 -  Find the average `total_acidity` for each of the `quality` values. 

6 - Find the top 5 `density` values. 

7 - Find the 10 lowest `sulphates` values. 

### Part 4 - Practice with Plotting 

1. Plot the average amount of `chlorides` for each `quality` value (1 from Part 3). 
2. Plot the `alcohol` values against `pH` values. Does there appear to be any relationship between the two?
3. Plot `total_acidity` values against `pH` values. Does there appear to be any relationship between the two?
4. Plot a histogram of the `quality` values. Are they evenly distributed within the data set?
5. Plot a boxplot to look at the distribution of `citric acid`. 

### Part 5 - Putting it All Together 

Now that you've worked on all the basics with one data set, it's time to do it with a second! This time, though, you'll go through the process of downloading the data set yourself. You'll also go through the process of learning to ask questions of the data (i.e. you won't be given any questions). We'll point you to a number of different data sets, and let you go at it. In reality, this is often how data science works. There isn't a clear-cut set of instructions on what to do - you kind of just dive into the data and see what you find!

Your goal by the end of `Part 5` is to be able to tell a story with your data. Whether that means you query it and find something interesting, examine a number of different columns and their values, or plot a couple of different columns, it doesn't matter. You should aim to find at least one piece of interesting information in your data (and ideally even more than one). Then, tell your peers and the instructors what you've found! 

Potential data sources: 

1. [Forest-fires](http://archive.ics.uci.edu/ml/datasets/Forest+Fires)
2. [Iris](http://archive.ics.uci.edu/ml/datasets/Iris)
3. [Another wine data set](http://archive.ics.uci.edu/ml/datasets/Wine)
4. [Abalone](http://archive.ics.uci.edu/ml/datasets/Abalone)
5. [Adult Income data set](http://archive.ics.uci.edu/ml/datasets/Adult)

The links above are all to the home pages of these data sources. At the top of these pages, you will find a link the the `Data Folder` where you can actually find the data. The majority of these data sets don't come in `.csv` format. While one of the datasets is available in `.csv` format, we encourage you to pick whatever data set you find most interesting (regardless of the format), and challenge yourself to read the necessary documentation and go through the process of figuring out how to get the data from the web and into a `DataFrame` (the instructors will also be around to help).